### PR TITLE
New Lambda packaging

### PR DESCRIPTION
**Pull request by patriot1burke**
* Extension now generates function.zip, sam.jvm.yaml, sam.native.yaml, manage.sh, and an example bootstrap script file
* uber-jar is no longer required
* Both JVM mode and Native mode generate a function.zip file. 
* lib/ directory is included in function.zip if uber-jar=false

### CODE REVIEW

**Comment by patriot1burke** [comment]
/cc 
@sherl0cks 
@Nxtra
@oztimpower

**Comment by sherl0cks** [comment]
@patriot1burke very interesting. Thanks for the tag Bill. WIll queue this up for tomorrow.

File: bom/deployment/pom.xml [hunk 1]
```diff
@@ -16,6 +16,14 @@
     <dependencyManagement>
         <dependencies>
 
+            <!-- Third party deployment dependencies -->
+            <dependency>
+                <groupId>org.apache.commons</groupId>
+                <artifactId>commons-compress</artifactId>
+                <version>1.20</version>
```
Comment (stuartwdouglas) [inline bom/deployment/pom.xml:23]: This should be a property

File: extensions/amazon-lambda-http/deployment/src/main/java/io/quarkus/amazon/lambda/http/deployment/AmazonLambdaHttpProcessor.java [hunk 2]
```diff
@@ -60,4 +63,23 @@ void setTempDir(BuildProducer<SystemPropertyBuildItem> systemProperty) {
         systemProperty.produce(new SystemPropertyBuildItem(FileResolver.CACHE_DIR_BASE_PROP_NAME, "/tmp"));
     }
 
+    @BuildStep
+    public void generateScripts(OutputTargetBuildItem target,
+            BuildProducer<ArtifactResultBuildItem> artifactResultProducer) throws Exception {
+        String lambdaName = LambdaUtil.artifactToLambda(target.getBaseName());
+
+        String output = LambdaUtil.copyResource("lambda/bootstrap-example.sh");
+        LambdaUtil.writeFile(target, "bootstrap-example.sh", output);
```
Comment (stuartwdouglas) [inline extensions/amazon-lambda-http/deployment/src/main/java/io/quarkus/amazon/lambda/http/deployment/AmazonLambdaHttpProcessor.java:72]: I can't help thinking that we need a build item to represent these additional files, so we have some way of tracking all the outputs of the build. I don't actually have a use case for this at the moment and this is fine for now, but maybe something we should revisit later.

File: extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaUtil.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public static void writeFile(OutputTargetBuildItem target, String name, String output) throws IOException {
+        FileOutputStream os = new FileOutputStream(target.getOutputDirectory().resolve(name).toFile());
+        os.write(output.getBytes(StandardCharsets.UTF_8));
+        os.close();
+    }
+
+    public static String copyResource(String resource) throws Exception {
+        InputStream inputStream = Thread.currentThread().getContextClassLoader().getResourceAsStream(resource);
```
Comment (stuartwdouglas) [inline extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaUtil.java:50]: This should be in a try with resources block.

File: extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaUtil.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            split = split.trim();
+            if (split.isEmpty())
+                continue;
+            name = name + split.substring(0, 1).toUpperCase() + split.substring(1).toLowerCase();
+        }
+        return name;
+    }
+
+    public static void writeFile(OutputTargetBuildItem target, String name, String output) throws IOException {
+        FileOutputStream os = new FileOutputStream(target.getOutputDirectory().resolve(name).toFile());
```
Comment (stuartwdouglas) [inline extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaUtil.java:44]: This should be in a try with resources block, or better yet just use Files.write()

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

File: extensions/amazon-lambda-http/deployment/src/main/resources/http/sam.jvm.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  Resources:
+    ${lambdaName}:
+      Type: AWS::Serverless::Function
+      Properties:
+        Handler: io.quarkus.amazon.lambda.runtime.QuarkusStreamHandler::handleRequest
+        Runtime: java8
+        CodeUri: function.zip
+        MemorySize: 512
+        Policies: AWSLambdaBasicExecutionRole
```
Comment (oztimpower) [inline extensions/amazon-lambda-http/deployment/src/main/resources/http/sam.jvm.yaml:18]: Update to java11 - consistent with other changes?

File: extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/FunctionZipProcessor.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+     * @param artifactResultProducer
+     * @param jar
+     * @throws Exception
+     */
+    @BuildStep(onlyIf = IsNormal.class, onlyIfNot = NativeBuild.class)
+    public void jvmZip(OutputTargetBuildItem target,
+            BuildProducer<ArtifactResultBuildItem> artifactResultProducer,
+            JarBuildItem jar) throws Exception {
+        log.info("**jvmZip");
+        Path zipPath = target.getOutputDirectory().resolve("function.zip");
```
Comment (oztimpower) [inline extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/FunctionZipProcessor.java:47]: logging in this method looks like it was used for debug

Comment (patriot1burke) [reply extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/FunctionZipProcessor.java:47]: yikes, lol!

File: extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaUtil.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+     *
+     * @param basename
+     * @return
+     */
+    public static String artifactToLambda(String basename) {
+        if (basename.endsWith("-SNAPSHOT"))
+            basename = basename.substring(0, basename.length() - "-SNAPSHOT".length());
+        String name = convertToken(basename, "[.0-9-]");
+        return name.trim();
+    }
```
Comment (oztimpower) [inline extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaUtil.java:27]: Should that be searching for any classifier, not just SNAPSHOT?

Comment (patriot1burke) [reply extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/LambdaUtil.java:27]: Other classifiers will be pruned if they are numeric.  I just don't have enough info on the artifact, and didn't want to change quarkus core to add it.

File: extensions/amazon-lambda/deployment/src/main/resources/lambda/manage.sh [hunk 8]
```diff
@@ -43,10 +43,10 @@ function cmd_update() {
     --zip-file ${ZIP_FILE}
 }
 
-FUNCTION_NAME=${resourceName}Function
+FUNCTION_NAME=${lambdaName}
 HANDLER=io.quarkus.amazon.lambda.runtime.QuarkusStreamHandler::handleRequest
-RUNTIME=java8
-ZIP_FILE=fileb://target/${artifactId}-${version}-runner.jar
+RUNTIME=java11
+ZIP_FILE=fileb://target/function.zip
```
Comment (oztimpower) [inline extensions/amazon-lambda/deployment/src/main/resources/lambda/manage.sh:50]: manage.sh still has target hardcoded, for both JVM and native file locations.

I cloned the branch from your repository and ran the maven archetype, and verified for gradle it is still showing as target (and not build).

File: docs/src/main/asciidoc/amazon-lambda-http.adoc [hunk 9]
```diff
@@ -191,16 +200,33 @@ Responses for binary types will be automatically encoded with base64.  This is d
...
+
 == Examine the POM
 
-If you want to adapt an existing RESTEasy, Undertow, or Vert.x Web project to Amazon Lambda, there's a couple
-of things you need to do.  Take a look at the generate example project to get an example of what you need to adapt.
+There is nothing special about the POM other than the inclusion of the `quarkus-amazon-lambda-http` extension
+as a dependencies.  The extension automatically generates everything you might need for your lambda deployment.
+
+NOTE: In previous versions of this extension you had to set up your pom or gradle
+to zip up your executable for native deployments, but this is not the case anymore.
```
Comment (sherl0cks) [inline docs/src/main/asciidoc/amazon-lambda-http.adoc:223]: This appears to be duplicated text from lines 205-209. Is that intentional?

File: docs/src/main/asciidoc/amazon-lambda.adoc [hunk 10]
```diff
@@ -423,16 +425,35 @@ NOTE: If using Gradle, the path to the binaries in the YAML templates must be ch
...
+
+There are times you may want to set a specific system properties or other arguments when lambda invokes
+your native quarkus lambda deployment.  If you include a `bootstrap` script file within
+`zip.native`, the quarkus extension will automatically rename the executable to `runner` within
+`function.zip` and set the unix mode of the `bootstrap` script to executable.
+
+NOTE: The native executable must be referenced as `runner` if you include a custom `bootstrap` script.
+
+The extension generates an example script within `target/bootstrap-example.sh`.
+
```
Comment (sherl0cks) [inline docs/src/main/asciidoc/amazon-lambda.adoc:456]: Great docs additions here.

**Comment by patriot1burke** [comment]
@gsmet @stuartwdouglas @geoand 

Can one of you approve?  The build failure was a timeout during non-lambda tests.

**Comment by geoand** [comment]
I'll leave merging to @stuartwdouglas since I have very little experience with the lambda stuff.

**Review by stuartwdouglas — APPROVED** [review]

**Comment by stuartwdouglas** [comment]
This looks awesome, it's a huge improvement over what we had

**Comment by geoand** [comment]
@stuartwdouglas @patriot1burke should we mark this as `noteworthy`?

**Comment by stuartwdouglas** [comment]
yes

**Comment by oztimpower** [comment]
@patriot1burke Similarly very excited and impressed by these improvements; looking forward to now including my build! A big thanks, makes things much simpler.

**Pull request merged**